pkg/handlers/v1/nexposevulnnotifier.go

```diff
@@ -0,0 +1,63 @@
+package v1
+
+import (
+	"context"
+
+	"github.com/asecurityteam/nexpose-vuln-notifier/pkg/domain"
+)
+
+// ScanInfo represents the incoming payload
+type ScanInfo struct {
+	// ScanID is the ID of the completed scan whose vulnerabilities we'll process
+	ScanID string `json:"scan_id"`
+	// SiteID is the ID of the site that just got scanned
+	SiteID string `json:"site_id"`
+}
+
+// NexposeVulnNotificationHandler is a lambda handler that fetches Nexpose Assets and sends them to an event stream
+type NexposeVulnNotificationHandler struct {
+	AssetFetcher domain.AssetFetcher
+	LogFn        domain.LogFn
+	StatFn       domain.StatFn
+}
+
+// Handle is an AWS Lambda handler that takes in a ScanID and SiteID for a Nexpose scan that has completed,
+// get all the Assets in the site that was scanned, hydrates the asset with the vulnerabilities that were found
+// on that asset and publishes the asset to a stream
+func (h *NexposeVulnNotificationHandler) Handle(ctx context.Context, in ScanInfo) {
+	logger := h.LogFn(ctx)
+
+	assetChan, errChan := h.AssetFetcher.FetchAssets(ctx, in.SiteID)
```

Comment:
Can you provide some addition context as to the intended flow of the program? `FetchAssets` is async, but we immediately block and wait for a response. What was the original intent for having `FetchAssets` async?

Somewhat related, it is _usually_ uncommon to see async functions in go for a [number of reasons](https://about.sourcegraph.com/go/idiomatic-go). One thing that would help is to move the goroutine outside of the function. e.g.

Reply:
Yes. AssetFetcher gets called here and will run and immediately return the assetChan and errorChan. Then the for/select/case block here will listen for assets and errors on the channels. Meanwhile, the `go func()` piece in FetchAssets will run and make its frist request to Nexpose to get the total number of pages it'll need to retrieve from Nexpose and spawn off that many goroutines. As the goroutines complete, they'll add assets to assetChan and mark the goroutine as done with `wg.Done` and the Handler will immediately receive the assets (or errors). The wg.Wait() at the end of the `go func()` will make sure all the goroutines have finished before closing both channels. We don't block on anything here so that as soon as a page of assets are returned from Nexpose we can immediately add them to the channel and the Handler can receive it to make another request to get the vulns for that asset without having to wait for all the other requests to get the assets to complete.

I understand and agree that async functions aren't usually needed, but when we need to get thousands of assets, we don't want to wait for all those requests to return before we can continue.

Reply:
Ah I see. I missed that in the comment.  FWIW, there is nothing wrong with doing that in the handler, however in the past we have tried to do those sorts of tasks inside decorators or other packages.

> Handler can receive it to make another request to get the vulns for that asset without having to wait for all the other requests to get the assets to complete

It does seem that in order to not be a bottleneck this will have to spin off yet another goroutine. Having the request to fetch the vulns be a decorator for the asset fetcher would allow you to do the work in the same same goroutine. However, I don't have enough context to know how the Nexpose API works, so I may be way off

Reply:
Here is an [example](https://bitbucket.org/hipchat/lastread/src/master/lib/lastread/querier.go).  The lastread Querier fans out requests to the Fetcher which are decorated with a retrier. In your case though, you could decorate them with a vuln fetcher.

Reply:
So after talking IRL, we decided that this nexpose-vuln-notifier should be split into 2 services, one that fetches assets from a site, and another one that adds the list of vulnerabilities to the asset and then gets all the vulnerability information for each vulnerability on the asset. As a result, here in nexposevulnnotifier.go, we'll call fetchAssets and publish the assets we receive onto a queue for the next service to consume.

This decorator pattern might be a good fit for the next ticket that will add publishing to the queue here.